receiver.py

```diff
@@ -72,13 +82,25 @@ def data_lines(self):
                         'source_addr_long', self.source_addr_long)
                 self.source_addr = packet.get(
                         'source_addr', self.source_addr)
+                self.rssi = packet.get('rssi', self.source_addr)
 
                 payload += packet['rf_data']
-                if x < self.expected_packets - 1 and len(payload) < 100:
-                    break;
 
-            # let our data be processed
-            yield self.data_shape.unpack(payload)
+                # Read first two bytes, to determine packet type
+                self.packet_type = int(payload[:2], 2)
+
+                #Preallocate space
+                stored_data = '' 
+                # Unpack Struct according to ID, and update global parameters
+                for i in xrange(len(self.data_shape)):
```

Comment:
Can just do `for data_shape in possible_data_shapes` instead of involving the len call + xrange.

Also, iirc, xrange == range in python 3.

Also, if you change it to be a dict, then you can do `for data_type, data_shape in self possible_data_shapes:`